tests/server.c

```diff
@@ -0,0 +1,359 @@
+  if(result != CRUSTLS_DEMO_OK) {
+    return NULL;
+  }
+
+  result = read_file(keyfile, keybuf, sizeof(keybuf), &keybuf_len);
+  if(result != CRUSTLS_DEMO_OK) {
+    return NULL;
+  }
+
+  const struct rustls_certified_key *certified_key;
+  result = rustls_certified_key_build(
+    (uint8_t*)certbuf, certbuf_len, (uint8_t*)keybuf, keybuf_len, &certified_key);
+  if(result != RUSTLS_RESULT_OK) {
+    print_error("parsing certificate and key", result);
+    return NULL;
+  }
+  return certified_key;
+}
+
+int
+main(int argc, const char **argv)
+{
+  int ret = 1;
+  int result = 1;
+  const char *certfile = argv[1];
+  const char *keyfile = argv[2];
+  struct rustls_server_config_builder *config_builder =
+    rustls_server_config_builder_new();
+  const struct rustls_server_config *server_config = NULL;
+  struct rustls_connection *rconn = NULL;
+
+  if(argc <= 2) {
+    fprintf(stderr,
+            "usage: %s cert.pem key.pem\n\n"
+            "Listen on port 8443 with the given cert and key.\n",
+            argv[0]);
+    goto cleanup;
+  }
+
+  const struct rustls_certified_key *certified_key =
+    load_cert_and_key(certfile, keyfile);
+  if(certified_key == NULL) {
+    goto cleanup;
+  }
+
+  rustls_server_config_builder_set_certified_keys(config_builder, &certified_key, 1);
+  server_config = rustls_server_config_builder_build(config_builder);
+
+#ifdef _WIN32
+  WSADATA wsa;
+  WSAStartup(MAKEWORD(1, 1), &wsa);
+#endif
+
+  int sockfd = socket(AF_INET, SOCK_STREAM, 0);
+  if(sockfd < 0) {
+    fprintf(stderr, "making socket: %s", strerror(errno));
+  }
+
+  struct sockaddr_in my_addr, peer_addr;
+  memset(&my_addr, 0, sizeof(struct sockaddr_in));
+  /* Clear structure */
+  my_addr.sin_family = AF_INET;
+  my_addr.sin_addr.s_addr = INADDR_ANY;
+  my_addr.sin_port = htons(8443);
+
+  if(bind(sockfd, (struct sockaddr *)&my_addr, sizeof(struct sockaddr_in)) ==
+     -1) {
+    perror("bind");
+    goto cleanup;
+  }
+
+  if(listen(sockfd, 50) == -1) {
+    perror("listen");
+    goto cleanup;
+  }
+  fprintf(stderr, "listening on localhost:8443\n");
+
+  while (true) {
+    socklen_t peer_addr_size;
+    peer_addr_size = sizeof(struct sockaddr_in);
+    int clientfd =
+      accept(sockfd, (struct sockaddr *)&peer_addr, &peer_addr_size);
+    if(clientfd < 0) {
+      perror("accept");
+      goto cleanup;
+    }
+
+    nonblock(clientfd);
+
+    result = rustls_server_connection_new(server_config, &rconn);
+    if(result != RUSTLS_RESULT_OK) {
+      print_error("making session", result);
+      goto cleanup;
+    }
+
+    pthread_t thrd;
+    conndata_t *conndata;
+    conndata = calloc(1, sizeof(conndata_t));
+    conndata->fd = clientfd;
+    conndata->rconn = rconn;
+    ret = pthread_create(&thrd, NULL, handle_conn, conndata);
+    if (ret != 0) {
+      fprintf(stderr, "error from pthread_create: %d\n", ret);
+      goto cleanup;
+    }
+    pthread_join(thrd, NULL);
```

Comment:
What's the point of handling connections on threads if you're going to block the loop that does accept on `pthread_join(3)` anyway? I think it's perfectly reasonable for this server to only handle one connection at a time, but you could just block on a call to `handle_conn` in the loop and make your life a little easier.